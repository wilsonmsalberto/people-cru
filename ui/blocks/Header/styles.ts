import styled from '@emotion/styled';

type StyledComponentProps = {
  color: string;
  bgColor: string;
};

export const StyledHeader = styled.header`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 2rem;
  transition: 1s color linear, 1s background-color linear;
  color: ${(props: StyledComponentProps): string => props.color};
  background-color: ${(props: StyledComponentProps): string => props.bgColor};
`;

export const ThemeSwitcherButton = styled.button`
  padding: 0.5rem;
  border-radius: 0.5rem;
  color: ${(props: StyledComponentProps): string => props.color};
  background-color: ${(props: StyledComponentProps): string => props.bgColor};

  svg {
    width: 3rem;
    height: 3rem;
    position: relative;
    top: 1px;
    fill: ${(props: StyledComponentProps): string => props.color};
  }
`;
