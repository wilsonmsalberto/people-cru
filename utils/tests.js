/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable react/prop-types */
import '@testing-library/jest-dom';
import * as React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import { Dark, Light } from 'ui/theme';

jest.mock('next/router', () => ({
    useRouter() {
        return {
            pathname: '/edit/$id',
            route   : '/edit/$id',
            query   : { id: 1 },
            asPath  : '/edit/1',
        };
    },
}));

const themes = {
    dark : Dark,
    light: Light,
};

export const themesList = Object.keys(themes);

const Providers = ({ children, theme = 'light' }) => (
    <ThemeProvider theme={ themes[theme] }>
        { children }
    </ThemeProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };