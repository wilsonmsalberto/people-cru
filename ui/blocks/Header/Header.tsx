import React from 'react';

import ThemeSwitcher from 'ui/blocks/ThemeSwitcher';
import User from 'ui/blocks/User';

import { InnerWrapper, StyledHeader } from './styles';

const Header = (): React.ReactElement => (
    <StyledHeader data-testid="header">
        <InnerWrapper>
            <ThemeSwitcher />
            <User />
        </InnerWrapper>
    </StyledHeader>
);

export default Header;
