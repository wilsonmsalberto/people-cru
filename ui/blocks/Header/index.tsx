import React, { useCallback, useEffect } from 'react';

import { useTheme } from 'emotion-theming';

import { StyledHeader, ThemeSwitcherButton } from './styles';

import { ContrastIcon } from 'ui/components/SvgIcons';

const Header = ({
  activeTheme,
  setTheme,
}: {
  activeTheme: string;
  setTheme: Function;
}): React.ReactElement => {
  const { secondary, primary, buttonBgColor, buttonTextColor } = useTheme();

  const applyTheme = useCallback(
    (theme: string): void => {
      setTheme(theme);
      localStorage.setItem('theme', theme);
    },
    [setTheme]
  );

  const handleThemeSwitch = useCallback(
    (theme: string): void => {
      if (theme === 'light') {
        applyTheme('dark');
      }

      if (theme === 'dark') {
        applyTheme('light');
      }
    },
    [applyTheme]
  );

  useEffect(() => {
    const savedTheme = (localStorage && localStorage.getItem('theme')) || '';

    if (savedTheme && savedTheme !== activeTheme) {
      applyTheme(savedTheme);
    } else {
      applyTheme(activeTheme);
    }
  }, [activeTheme, applyTheme]);

  return (
    <StyledHeader bgColor={secondary} color={primary}>
      <ThemeSwitcherButton
        bgColor={buttonBgColor}
        color={buttonTextColor}
        onClick={(): void => handleThemeSwitch(activeTheme)}
      >
        <ContrastIcon />
      </ThemeSwitcherButton>
    </StyledHeader>
  );
};

const withContextHeader = (props: Record<string, any>): React.ReactElement => (
  <Header activeTheme={props.activeTheme} setTheme={props.setTheme} {...props} />
);

export default withContextHeader;
