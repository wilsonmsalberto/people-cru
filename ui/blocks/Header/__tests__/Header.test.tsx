import { render, themesList } from 'utils/tests';

// Components
import Header from '../Header';

const getRenderElement = (props: any) => render(
    <Header { ...props } />
);

describe('<Header />', () => {
    describe('Theme rendering: ', () => {
        test.each(themesList)('%p theme',
            (theme) => {
                const { getByTestId } = getRenderElement({ theme });

                const component = getByTestId('header');

                expect(component).toMatchSnapshot();
            }
        );
    });

    describe('Basic rendering: ', () => {
        it('should render properly', () => {
            const { getByTestId } = getRenderElement({});

            const component = getByTestId('header');

            expect(component).toBeVisible();
        });
    });
});
