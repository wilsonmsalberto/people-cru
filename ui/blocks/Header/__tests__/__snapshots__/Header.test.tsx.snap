// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Theme rendering:  "dark" theme 1`] = `
.emotion-0 {
  width: 100%;
  height: 8rem;
  padding: 2rem;
  -webkit-transition: 1s color linear,1s background-color linear;
  transition: 1s color linear,1s background-color linear;
  color: ;
  background-color: ;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
  max-width: 107.8rem;
  margin: 0 auto;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 4rem;
  height: 4rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  color: #624DE3;
  background-color: transparent;
}

.emotion-4 svg {
  position: relative;
  top: 0.1rem;
  -webkit-transition: 0.2s -webkit-transform linear,0.2s opacity linear;
  transition: 0.2s transform linear,0.2s opacity linear;
}

.emotion-4:hover svg {
  -webkit-transform: scale(1.1);
  -moz-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
  opacity: 0.85;
}

.emotion-6 {
  display: grid;
  grid-template-columns: 1fr auto;
  grid-template-rows: 1fr 1fr;
  grid-column-gap: 1.3rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
}

.emotion-8 {
  grid-column: 1;
  grid-row: span 2;
  width: 3.5rem;
  height: 3.5rem;
  border-radius: 50%;
  background-color: #FFE8DF;
}

.emotion-10 {
  font-weight: 500;
  font-size: 1.6rem;
  line-height: 1.9rem;
  color: #00234B;
}

.emotion-12 {
  font-weight: 500;
  font-size: 1.3rem;
  line-height: 1.6rem;
  color: #778CA3;
}

<header
  class="emotion-0 emotion-1"
  data-testid="header"
>
  <header
    class="emotion-2 emotion-3"
  >
    <button
      class="emotion-4 emotion-5"
      data-testid="themeswitcher"
    >
      <svg
        fill="currentColor"
        height="100%"
        viewBox="0 0 512 512"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Contrast Icon
        </title>
        <path
          d="M256 0C114.848 0 0 114.848 0 256s114.848 256 256 256 256-114.848 256-256S397.152 0 256 0zm0 480C132.48 480 32 379.52 32 256S132.48 32 256 32s224 100.48 224 224-100.48 224-224 224z"
        />
        <path
          d="M256 64v384c112 0 192-85.968 192-192 0-106.048-80-192-192-192z"
        />
      </svg>
    </button>
    <aside
      class="emotion-6 emotion-7"
      data-testid="user"
    >
      <div
        class="emotion-8 emotion-9"
      />
      <h2
        class="emotion-10 emotion-11"
      >
        Julie Howard
      </h2>
      <h3
        class="emotion-12 emotion-13"
      >
        Admin
      </h3>
    </aside>
  </header>
</header>
`;

exports[`<Header /> Theme rendering:  "light" theme 1`] = `
.emotion-0 {
  width: 100%;
  height: 8rem;
  padding: 2rem;
  -webkit-transition: 1s color linear,1s background-color linear;
  transition: 1s color linear,1s background-color linear;
  color: ;
  background-color: ;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
  max-width: 107.8rem;
  margin: 0 auto;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 4rem;
  height: 4rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  color: #624DE3;
  background-color: transparent;
}

.emotion-4 svg {
  position: relative;
  top: 0.1rem;
  -webkit-transition: 0.2s -webkit-transform linear,0.2s opacity linear;
  transition: 0.2s transform linear,0.2s opacity linear;
}

.emotion-4:hover svg {
  -webkit-transform: scale(1.1);
  -moz-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
  opacity: 0.85;
}

.emotion-6 {
  display: grid;
  grid-template-columns: 1fr auto;
  grid-template-rows: 1fr 1fr;
  grid-column-gap: 1.3rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
}

.emotion-8 {
  grid-column: 1;
  grid-row: span 2;
  width: 3.5rem;
  height: 3.5rem;
  border-radius: 50%;
  background-color: #FFE8DF;
}

.emotion-10 {
  font-weight: 500;
  font-size: 1.6rem;
  line-height: 1.9rem;
  color: #00234B;
}

.emotion-12 {
  font-weight: 500;
  font-size: 1.3rem;
  line-height: 1.6rem;
  color: #778CA3;
}

<header
  class="emotion-0 emotion-1"
  data-testid="header"
>
  <header
    class="emotion-2 emotion-3"
  >
    <button
      class="emotion-4 emotion-5"
      data-testid="themeswitcher"
    >
      <svg
        fill="currentColor"
        height="100%"
        viewBox="0 0 512 512"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Contrast Icon
        </title>
        <path
          d="M256 0C114.848 0 0 114.848 0 256s114.848 256 256 256 256-114.848 256-256S397.152 0 256 0zm0 480C132.48 480 32 379.52 32 256S132.48 32 256 32s224 100.48 224 224-100.48 224-224 224z"
        />
        <path
          d="M256 64v384c112 0 192-85.968 192-192 0-106.048-80-192-192-192z"
        />
      </svg>
    </button>
    <aside
      class="emotion-6 emotion-7"
      data-testid="user"
    >
      <div
        class="emotion-8 emotion-9"
      />
      <h2
        class="emotion-10 emotion-11"
      >
        Julie Howard
      </h2>
      <h3
        class="emotion-12 emotion-13"
      >
        Admin
      </h3>
    </aside>
  </header>
</header>
`;
